extendr-api/src/wrapper/macros.rs

```diff
@@ -0,0 +1,154 @@
+                #[doc = "   assert_eq!(vec.is_" $r_type:lower "(), true);"]
+                #[doc = "   assert_eq!(vec.len(), 10);"]
+                #[doc = "}"]
+                #[doc = "```"]
+                pub fn new(len: usize) -> $type {
+                    // TODO: Check if impacts performance.
+                    let iter = (0..len).map(|_| <$type_prim>::default());
+                    <$type>::from_values(iter)
+                }
+            }
+            paste::paste!{
+                #[doc = "Wrapper for creating ALTREP " $doc_name " (" $sexp ") vectors from iterators."]
+                #[doc = "The iterator must be exact, cloneable and implement Debug."]
+                #[doc = "If you want a more generalised constructor, use `iter.collect::<" $type ">()`."]
+                pub fn from_values<V>(values: V) -> Self
+                where
+                    V: IntoIterator,
+                    V::IntoIter: ExactSizeIterator + std::fmt::Debug + Clone + 'static + std::any::Any,
+                    V::Item: Into<$type_prim>,
+                {
+                    single_threaded(|| {
+                        let values: V::IntoIter = values.into_iter();
+
+                        let robj = if values.len() >= SHORT_VECTOR_LENGTH {
+                            Altrep::[<make_alt $r_type:lower _from_iterator>](values)
+                                .try_into()
+                                .unwrap()
+                        } else {
+                            let mut robj = Robj::alloc_vector($sexp, values.len());
+                            let dest: &mut [$type_prim] = robj.as_typed_slice_mut().unwrap();
+
+                            for (d, v) in dest.iter_mut().zip(values) {
+                                *d = v.into();
+                            }
+                            robj
+                        };
+                        Self { robj }
+                    })
+                }
+            }
+
+            paste::paste! {
+                #[doc = "Get a single element from the vector."]
+                #[doc = "Note that this is very inefficient in a tight loop."]
+                #[doc = "```"]
+                #[doc = "use extendr_api::prelude::*;"]
+                #[doc = "test! {"]
+                #[doc = "   let vec = " $type "::new(10);"]
+                #[doc = "   assert_eq!(vec.elt(0), <"$type_elem">::default());"]
+                #[doc = "   assert_eq!(vec.elt(9), <"$type_elem">::default());"]
+                #[doc = "   assert!(vec.elt(10).is_na());"]
+                #[doc = "}"]
+                #[doc = "```"]
+                pub fn elt(&self, index: usize) -> $type_elem {
+                    // Defensive check for oob
+                    if(index >= self.len()) {
+                        <$type_elem>::na()
+                    } else {
+                        unsafe { [<$r_type _ELT>](self.get(), index as R_xlen_t).into() }
+                    }
```

Comment:
`*_ELT` will check the bounds as of the next major release of R; I mean, the internal, macro version of `*_ELT` don't, but the function version of them do, and the internal version won't be available anymore. 

c.f. https://twitter.com/dvaughan32/status/1417647704289824770

<details>

<summary>e.g. function version of `INTEGER_ELT`</summary>

https://github.com/wch/r-source/blob/77d6c4371eaa690f151f36d5767be892e8a7b334/src/include/Rinlinedfuns.h#L393-L397

CHECK_VECTOR_INT_ELT
https://github.com/wch/r-source/blob/77d6c4371eaa690f151f36d5767be892e8a7b334/src/include/Rinlinedfuns.h#L277-L282

CHECK_BOUNDS_ELT
https://github.com/wch/r-source/blob/77d6c4371eaa690f151f36d5767be892e8a7b334/src/include/Rinlinedfuns.h#L266-L269

</details>

So, if we want to avoid the double check, we might want to access to the underlying pointer directly in the case of non-ALTREP, just like what cpp11 does.

https://github.com/r-lib/cpp11/blob/1b4beb25f56c8dca6e0448d5dc55fb64fddab5d1/inst/include/cpp11/integers.hpp#L29-L32

Reply:
I agree, but I did not change the functions that were already used in the case of `Integers`.
I agree that your solution is better, so perhaps you can make a follow-up PR to fix this?
This PR already covers a lot, I do not want to make more changes here.

Reply:
> I did not change the functions that were already used in the case of `Integers`.

I commented this because you added the bound check, but, okay, I don't intend to block the PR for this. This comment was just in case you didn't notice my reply on Discord. Let's discuss later :)